aidial_adapter_openai/utils/streaming.py

```diff
@@ -106,39 +109,26 @@ async def generate_stream(
         return
 
     if not stream_finished:
-        if last_chunk is not None:
-            logger.warning("Didn't receive chunk with the finish reason")
-
-            completion_tokens = tokenizer.calculate_tokens(total_content)
-            last_chunk["usage"] = {
-                "completion_tokens": completion_tokens,
-                "prompt_tokens": prompt_tokens,
-                "total_tokens": prompt_tokens + completion_tokens,
-            }
-            last_chunk["choices"][0]["delta"]["content"] = ""
-            last_chunk["choices"][0]["finish_reason"] = "length"
-
-            yield last_chunk
-        else:
+        if last_chunk is None:
             logger.warning("Received 0 chunks")
 
-            id = generate_id()
-            created = str(int(time()))
-            is_stream = True
-
-            yield build_chunk(
-                id,
-                "length",
-                {},
-                created,
-                is_stream,
+            last_chunk = build_chunk(
+                id=generate_id(),
                 model=deployment,
-                usage={
-                    "completion_tokens": 0,
-                    "prompt_tokens": prompt_tokens,
-                    "total_tokens": prompt_tokens,
-                },
+                is_stream=True,
+                created=str(int(time())),
+                message={},
+                finish_reason=None,
             )
+        else:
+            logger.warning("Didn't receive chunk with the finish reason")
+
+        finalize_last_chunk(last_chunk)
+
+        last_chunk["choices"][0]["delta"] = {}
```

Comment:
Last chunk could be some content, so it would be better to make
```python
last_chunk["choices"][0]["delta"] = last_chunk["choices"][0].get('delta') or {}
```

Reply:
Also, is there any case, when choices would be just empty array?

Reply:
The code was a bit confusing indeed. I've refactored it.

Empty array of choices does appear in cases when the model has nothing to report in the content of choices, but reports some auxilary information like usage.

Let's handle the case with multipe choices in a separate PR: https://github.com/epam/ai-dial-adapter-openai/issues/140